docs/book/how-to/handle-data-artifacts/datasets.md

```diff
@@ -0,0 +1,246 @@
+        )
+        dataset.read_data()
+        return dataset
+
+    def save(self, bq_dataset: BigQueryDataset) -> None:
+        metadata = {
+            "table_id": bq_dataset.table_id,
+            "project": bq_dataset.project,
+        }
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "w") as f:
+            json.dump(metadata, f)
+        if bq_dataset.df is not None:
+            bq_dataset.write_data()
+```
+
+## Managing Complexity in Pipelines with Multiple Data Sources
+
+When working with multiple data sources, it's crucial to design flexible pipelines that can handle different scenarios. Here's an example of how to structure a pipeline that works with both CSV and BigQuery datasets:
+
+```python
+from zenml import step, pipeline
+from typing_extensions import Annotated
+
+@step
+def extract_data_local(data_path: str = "data/raw_data.csv") -> CSVDataset:
+    return CSVDataset(data_path)
+
+@step
+def extract_data_remote(table_id: str) -> BigQueryDataset:
+    return BigQueryDataset(table_id)
+
+@step
+def transform_csv(dataset: CSVDataset) -> Annotated[CSVDataset, "transformed_dataset"]:
+    df = dataset.read_data()
+    # Transform data
+    transformed_df = df.copy()  # Apply transformations here
+    return CSVDataset("tmp/transformed_data.csv", df=transformed_df)
+
+@step
+def transform_bq(dataset: BigQueryDataset) -> Annotated[BigQueryDataset, "transformed_dataset"]:
+    df = dataset.read_data()
+    # Transform data
+    transformed_df = df.copy()  # Apply transformations here
+    return BigQueryDataset(table_id="project.dataset.transformed_table", df=transformed_df)
+
+@pipeline
+def etl_pipeline(mode: str = "develop"):
+    if mode == "develop":
+        raw_data = extract_data_local()
+        transformed_data = transform_csv(raw_data)
+    else:
+        raw_data = extract_data_remote(table_id="project.dataset.raw_table")
+        transformed_data = transform_bq(raw_data)
+```
+
+## Best Practices for Designing Flexible and Maintainable Pipelines
+
+When working with custom Dataset classes in ZenML pipelines, it's crucial to design your pipelines
+to accommodate various data sources and processing requirements.
+
+Here are some best practices to ensure your pipelines remain flexible and maintainable:
+
+1. **Use a common base class**: The `Dataset` base class allows for consistent handling of different data sources within your pipeline steps. This abstraction enables you to swap out data sources without changing the overall pipeline structure.
+
+```python
+@step
+def process_data(dataset: Dataset) -> ProcessedData:
+    data = dataset.read_data()
+    # Process data...
+    return processed_data
+```
+
+2. **Create specialized steps**: Implement separate steps for different dataset types to handle specific processing requirements while keeping your code modular. This approach allows you to tailor your processing to the unique characteristics of each data source.
```

Comment:
Isn't this the opposite of the first best practice?

Reply:
took it out